extensions/standard-processors/tests/unit/ProcessorTests.cpp

```diff
@@ -615,65 +615,131 @@ TEST_CASE("TestRPGValid", "[TestRPG6]") {
 
-    REQUIRE_FALSE(processor->isWorkAvailable());
+    REQUIRE_FALSE(processor_->isWorkAvailable());
   }
 
   SECTION("If there is both a penalized and a non-penalized flow file queued, there is work available") {
-    const auto normal_flow_file = session->create();
-    incoming_connection->put(normal_flow_file);
+    const auto normal_flow_file = session_->create();
+    incoming_connection_->put(normal_flow_file);
 
-    const auto penalized_flow_file = session->create();
-    session->penalize(penalized_flow_file);
-    incoming_connection->put(penalized_flow_file);
+    const auto penalized_flow_file = session_->create();
+    session_->penalize(penalized_flow_file);
+    incoming_connection_->put(penalized_flow_file);
 
-    REQUIRE(processor->isWorkAvailable());
+    REQUIRE(processor_->isWorkAvailable());
   }
 
   SECTION("When a penalized flow file is queued, there is work available after the penalty expires") {
-    processor->setPenalizationPeriodMsec(10);
+    processor_->setPenalizationPeriod(std::chrono::milliseconds{10});
 
-    const auto flow_file = session->create();
-    session->penalize(flow_file);
-    incoming_connection->put(flow_file);
+    const auto flow_file = session_->create();
+    session_->penalize(flow_file);
+    incoming_connection_->put(flow_file);
 
-    REQUIRE_FALSE(processor->isWorkAvailable());
+    REQUIRE_FALSE(processor_->isWorkAvailable());
     const auto penalty_has_expired = [flow_file] { return !flow_file->isPenalized(); };
     REQUIRE(utils::verifyEventHappenedInPollTime(std::chrono::seconds{1}, penalty_has_expired, std::chrono::milliseconds{10}));
-    REQUIRE(processor->isWorkAvailable());
+    REQUIRE(processor_->isWorkAvailable());
   }
 }
+
+TEST_CASE_METHOD(ProcessorWithIncomingConnectionTest, "A failed and re-penalized flow file does not block the incoming queue of the Processor", "[penalize]") {
+  processor_->setPenalizationPeriod(std::chrono::milliseconds{100});
+  const auto penalized_flow_file = session_->create();
+  session_->penalize(penalized_flow_file);  // first penalty duration is 100 ms
+  incoming_connection_->put(penalized_flow_file);
+  const auto penalty_has_expired = [penalized_flow_file] { return !penalized_flow_file->isPenalized(); };
+  REQUIRE(utils::verifyEventHappenedInPollTime(std::chrono::seconds{1}, penalty_has_expired, std::chrono::milliseconds{10}));
+
+  const auto flow_file_1 = session_->create();
+  incoming_connection_->put(flow_file_1);
+  const auto flow_file_2 = session_->create();
+  incoming_connection_->put(flow_file_2);
+  const auto flow_file_3 = session_->create();
+  incoming_connection_->put(flow_file_3);
+
+  REQUIRE(incoming_connection_->isWorkAvailable());
+  std::set<std::shared_ptr<core::FlowFile>> expired_flow_files;
+  const auto next_flow_file_1 = incoming_connection_->poll(expired_flow_files);
+  REQUIRE(next_flow_file_1 == penalized_flow_file);
+
+  session_->penalize(penalized_flow_file);  // second penalty duration is 200 ms
```

Comment:
does this comment apply?

Reply:
Nope.  Removed in https://github.com/apache/nifi-minifi-cpp/pull/1027/commits/a42c324123487d8a05be45c236ef1b42fd24457e.